src/coreclr/jit/codegenxarch.cpp

```diff
@@ -11188,12 +11202,25 @@ void CodeGen::genZeroInitFrameUsingBlockInit(int untrLclHi, int untrLclLo, regNu
 //             funclet frames: this will be FuncletInfo.fiSpDelta.
 void CodeGen::genPreserveCalleeSavedFltRegs(unsigned lclFrameSize)
 {
-    genVzeroupperIfNeeded(false);
     regMaskTP regMask = compiler->compCalleeFPRegsSavedMask;
 
     // Only callee saved floating point registers should be in regMask
     assert((regMask & RBM_FLT_CALLEE_SAVED) == regMask);
 
+    if (GetEmitter()->Contains256bitOrMoreAVX())
+    {
+        // The Intel optimization manual guidance in `3.11.5.3 Fixing Instruction Slowdowns` states:
+        //   Insert a VZEROUPPER to tell the hardware that the state of the higher registers is clean
+        //   between the VEX and the legacy SSE instructions. Often the best way to do this is to insert a
+        //   VZEROUPPER before returning from any function that uses VEX (that does not produce a VEX
+        //   register) and before any call to an unknown function.
+
+        // Since this function used 256-bit or more AVX, we need to ensure we emit a vzeroupper.
+        // TODO-CQ: We should be able to check the return type and skip this if it is TYP_SIMD32 or TYP_SIMD64
+
+        instGen(INS_vzeroupper);
+    }
```

Comment:
Given the rules in the post, one might think this is unnecessary. However, there are two interesting cases to consider:
1. A method marked `[UnmanagedCallersOnly]` where we know we are transferring control back to native (therefore are `managed to unmanaged`
2. A method that was called from a R2R method which was not compiled for AVX+ (in which case the managed code might be executing legacy encoded floating-point instructions).

Reply:
We can of course skip this if the method return type or any of its parameters are `Vector<T>`, `Vector256<T>`, or `Vector512<T>`. But, there didn't seem to be a trivial way to check that today.